shared/src/main/scala/mlscript/TyperDatatypes.scala

```diff
@@ -110,8 +110,19 @@ abstract class TyperDatatypes extends TyperHelpers { self: Typer =>
       if (fields.isEmpty) ExtrType(false)(prov) else RecordType(fields)(prov)
   }
   
-  case class TupleType(fields: List[Opt[Var] -> SimpleType])(val prov: TypeProvenance) extends MiscBaseType {
+  sealed abstract class ArrayBase extends MiscBaseType {
+    def inner: SimpleType
+  }
+
+  case class ArrayType(val inner: SimpleType)(val prov: TypeProvenance) extends ArrayBase {
+    lazy val level: Int = inner.level
```

Comment:
This does not need to be a `lazy val` as the computation is trivial and needs not be stored.